### PR TITLE
add color to category

### PR DESCRIPTION
**Pull request by ADI-projects30**
1. Create front for insert category with color
2. In add event: will be show all user categories with their color

### CODE REVIEW

File: app/routers/categories.py [hunk 1]
```diff
@@ -87,3 +124,38 @@ def get_user_categories(db_session: Session,
...
         return categories
+
+
+@router.post("/for_categories_test")
+async def for_category_test(request: Request,
+                            category_name: str = Form(None),
+                            chosen_color: str = Form(None),
+                            db_sess: Session = Depends(get_test_db)):
+    """
+    This route is only for run tests, user can't know this url and even
```
Comment (ivarshav) [inline app/routers/categories.py]: Consider: ...,user can't access this url - will get Method Not Allowed status

File: app/routers/categories.py [hunk 2]
```diff
@@ -43,22 +45,57 @@ def get_categories(request: Request,
                                    f"unallowed params.")
 
 
+@router.get("/event/edit")
```
Comment (ivarshav) [inline app/routers/categories.py]: Maybe it should be in events router (in `events.py` file) and not in categories router?
or maybe the routing should start with `categories` since it's related to categories?

Generally a router file (for example this one - categories.py) should contain apis with same prefix in my opinion.

File: app/routers/categories.py [hunk 3]
```diff
@@ -43,22 +45,57 @@ def get_categories(request: Request,
...
+    })
+
+
 # TODO(issue#29): get current user_id from session
-@router.post("/")
-async def set_category(category: CategoryModel,
-                       db_sess: Session = Depends(get_db)) -> Dict[str, Any]:
+@router.post("/categories")
+async def set_category(request: Request,
+                       category_name: str = Form(None),
```
Comment (ivarshav) [inline app/routers/categories.py:72]: You should send from FE a post request with body which includes a categoryModel == a json (dict).
Doesn't it work?

File: app/routers/categories.py [hunk 4]
```diff
@@ -43,22 +45,57 @@ def get_categories(request: Request,
...
+                        color=chosen_color,
+                        user_id=user_id)
     except IntegrityError:
         db_sess.rollback()
-        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
-                            detail=f"category is already exists for "
-                                   f"user {category.user_id}.")
-    else:
-        return {"category": cat.to_dict()}
+        message = "Category is already exists"
```
Comment (ivarshav) [inline app/routers/categories.py]: Did you change the message by mistake?
The original was "category is already exists for user {user_id}."

File: app/routers/categories.py [hunk 5]
```diff
@@ -43,22 +45,57 @@ def get_categories(request: Request,
...
+        return templates.TemplateResponse("categories.html", {
+            "request": request,
+            "message": message,
+            "category_name": category_name,
+            "chosen_color": chosen_color,
+        })
+    message = f"Congratulation! You have created the category: {category_name}"
```
Comment (ivarshav) [inline app/routers/categories.py]: Consider: You have created a new category: ..

File: app/routers/categories.py [hunk 6]
```diff
@@ -32,7 +34,7 @@ class Config:
 
 
 # TODO(issue#29): get current user_id from session
-@router.get("/")
+@router.get("/categories/by_parameters")
```
Comment (ivarshav) [inline app/routers/categories.py]: I wonder.. if there are 2 get requests with different params (this one and `@router.get("/categories")`, 
can't they have the same route?

File: app/routers/categories.py [hunk 7]
```diff
@@ -43,22 +45,57 @@ def get_categories(request: Request,
                                    f"unallowed params.")
 
 
+@router.get("/event/edit")
+def your_categories(request: Request,
+                    db_session: Session = Depends(get_db)) -> List[Category]:
+    """
+    will add all user categories to dropdown list
```
Comment (ivarshav) [inline app/routers/categories.py]: """Return all categories created by user."""

File: app/routers/categories.py [hunk 8]
```diff
@@ -43,22 +45,57 @@ def get_categories(request: Request,
                                    f"unallowed params.")
 
 
+@router.get("/event/edit")
+def your_categories(request: Request,
```
Comment (ivarshav) [inline app/routers/categories.py]: What the request can include?

File: app/routers/categories.py [hunk 9]
```diff
@@ -43,22 +45,57 @@ def get_categories(request: Request,
...
+                    db_session: Session = Depends(get_db)) -> List[Category]:
+    """
+    will add all user categories to dropdown list
+    """
+    user_id = 1
+    categories_list = get_user_categories(db_session, user_id)
```
Comment (ivarshav) [inline app/routers/categories.py]: consider using `if validate_request_params(request.query_params):` before if needed.
Please note that right now this function expects params names: "name" and "color", so you might need to change it or keep aligned with current names.

File: app/routers/categories.py [hunk 10]
```diff
@@ -43,22 +45,57 @@ def get_categories(request: Request,
                                    f"unallowed params.")
 
 
+@router.get("/event/edit")
+def your_categories(request: Request,
+                    db_session: Session = Depends(get_db)) -> List[Category]:
+    """
+    will add all user categories to dropdown list
+    """
+    user_id = 1
```
Comment (ivarshav) [inline app/routers/categories.py]: Please add TODO

File: app/routers/categories.py [hunk 11]
```diff
@@ -87,3 +124,38 @@ def get_user_categories(db_session: Session,
         return []
     else:
         return categories
+
+
+@router.post("/for_categories_test")
+async def for_category_test(request: Request,
```
Comment (ivarshav) [inline app/routers/categories.py]: Not sure why you need this? it looks very similar the other function above

File: app/static/categories_style.css [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+  text-align: center;
+}
+
+input {
+  border-radius: 0.5em;
+  color: plum;
+  padding: 0.5em;
+}
+input[type="color"].custom {
+      padding: 0;
```
Comment (ivarshav) [inline app/static/categories_style.css]: keep indentation

File: app/static/categories_style.css [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    border: none;
+    height: 30px;
```
Comment (ivarshav) [inline app/static/categories_style.css]: Prefer not to use `px`

File: app/routers/categories.py (hunk 5 above)
Comment (IdanPelled) [reply app/routers/categories.py]: +1

File: app/templates/categories.html [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+<link
+      rel="stylesheet"
+      href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css"
+      integrity="sha384-JcKb8q3iqJ61gNV9KGb8thSsNjpSL0n8PARn9HuZOnIxN0hoP+VmmDGMN5t9UJ0Z"
+      crossorigin="anonymous"
+    />
+{% endblock %}
+
+{% block content %}
+<div class="container mt-4">
```
Comment (ivarshav) [inline app/templates/categories.html:10]: Add indentation please

File: app/templates/categories.html [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+<h1>It's time to make some decisions</h1>
+<h1>
+  Here you can create your unique categories and choose your favorite color
```
Comment (ivarshav) [inline app/templates/categories.html:13]: Add `.` or `:` in the end of the sentence

File: app/routers/categories.py [hunk 16]
```diff
@@ -87,3 +124,38 @@ def get_user_categories(db_session: Session,
...
+    message = ""
+    user_id = 1
+    try:
+        Category.create(db_sess,
+                        name=category_name,
+                        color=chosen_color,
+                        user_id=user_id)
+    except IntegrityError:
+        db_sess.rollback()
+        message = "category is already exists"
```
Comment (IdanPelled) [inline app/routers/categories.py]: Without "is"

File: tests/test_categories.py [hunk 17]
```diff
@@ -9,31 +10,31 @@
...
-                                                     "name": "Guitar Lesson",
-                                                     "color": "121212"})
-        assert response.status_code == status.HTTP_400_BAD_REQUEST
-        assert TestCategories.CATEGORY_ALREADY_EXISTS_MSG in \
-               response.json()["detail"]
+        response = client.post("/for_categories_test",
+                               data={"category_name": "Guitar Lesson",
+                                     "chosen_color": "121212",
+                                     "user_id": sender.id})
+        assert response.status_code == status.HTTP_200_OK
```
Comment (ivarshav) [inline tests/test_categories.py]: consider using response.ok

File: tests/test_categories.py [hunk 18]
```diff
@@ -114,3 +117,8 @@ def raise_error(param):
 
         session.query = mock.Mock(side_effect=raise_error)
         assert get_user_categories(session, 1) == []
+
+    @staticmethod
+    def test_event_status_ok(client):
```
Comment (ivarshav) [inline tests/test_categories.py]: Does the test check creating a new event with category?
Please rename or add doc.

What about bad status?

File: app/templates/event/partials/edit_event_details_tab.html [hunk 19]
```diff
@@ -33,7 +33,19 @@
             <option value="blue">Blue</option>
         </select>
     </div>
-
+    <div class="user_categories">
```
Comment (ivarshav) [inline app/templates/event/partials/edit_event_details_tab.html]: Is that for creating a new event or also edit event?

Is it possible to edit a category after creating it?

File: app/routers/categories.py [hunk 20]
```diff
@@ -1,19 +1,21 @@
...
+
 
-router = APIRouter(
-    prefix="/categories",
-    tags=["categories"],
-)
+router = APIRouter()
 
 
 class CategoryModel(BaseModel):
```
Comment (ivarshav) [inline app/routers/categories.py]: Does someone use it now? (since you don't use it in the post requset)

File: app/routers/categories.py (hunk 1 above)
Comment (IdanPelled) [reply app/routers/categories.py]: First of all, this is an open source project, so the user **can** know the route.
Secondly, the user can access the route with an external post request.
[Here](https://stackoverflow.com/questions/10637951/how-to-block-external-http-requests-securing-ajax-calls/10769466#10769466) you can see that it's very hard to prevent faking a post request.

I recommend that you find a better way implement this function.

File: app/routers/categories.py [hunk 21]
```diff
@@ -87,3 +124,38 @@ def get_user_categories(db_session: Session,
...
+            "category_name": category_name,
+            "chosen_color": chosen_color,
+        })
+    message = f"Congratulation! You have created the category {category_name}"
+    return templates.TemplateResponse("categories.html", {
+        "request": request,
+        "message": message,
+        "category_name": category_name,
+        "chosen_color": chosen_color,
+    })
```
Comment (IdanPelled) [inline app/routers/categories.py]: This part is repeating itself, consider fixing it

File: tests/test_categories.py [hunk 22]
```diff
@@ -69,18 +72,18 @@ def test_get_category_by_name(client, sender, category):
...
+        response = client.get(f"/categories/by_parameters/?"
+                              f"user_id={category.user_id}&"
                               f"color={category.color}")
         assert response.ok
         assert set(response.json()[0].items()) == {
             ("user_id", category.user_id), ("color", "121212"),
             ("name", "Guitar Lesson"), ("id", category.id)}
 
     @staticmethod
-    def test_get_category_bad_request(client):
```
Comment (ivarshav) [inline tests/test_categories.py:94]: Why did you remove this one?

File: tests/test_categories.py [hunk 23]
```diff
@@ -69,18 +72,18 @@ def test_get_category_by_name(client, sender, category):
...
-        response = client.get("/categories/")
-        assert response.status_code == status.HTTP_400_BAD_REQUEST
-        assert TestCategories.UNALLOWED_PARAMS in response.json()["detail"]
+    def test_get_category_ok_request(client):
+        response = client.get("/categories")
+        assert response.status_code == status.HTTP_200_OK
```
Comment (ivarshav) [inline tests/test_categories.py]: consider using response.ok

File: tests/test_categories.py [hunk 24]
```diff
@@ -9,31 +10,31 @@
 
 
 class TestCategories:
-    CATEGORY_ALREADY_EXISTS_MSG = "category is already exists for"
+    CATEGORY_ALREADY_EXISTS_MSG = b"category is already exists"
+    CREATE_CATEGORY = b"You have created"
     UNALLOWED_PARAMS = "contains unallowed params"
```
Comment (ivarshav) [inline tests/test_categories.py:17]: If you add `b` in the prefix of other string...maybe this one needs too?

File: app/routers/categories.py [hunk 25]
```diff
@@ -43,22 +45,57 @@ def get_categories(request: Request,
...
+
+
 # TODO(issue#29): get current user_id from session
-@router.post("/")
-async def set_category(category: CategoryModel,
-                       db_sess: Session = Depends(get_db)) -> Dict[str, Any]:
+@router.post("/categories")
+async def set_category(request: Request,
+                       category_name: str = Form(None),
+                       chosen_color: str = Form(None),
```
Comment (ivarshav) [inline app/routers/categories.py:73]: Where do you validate the color format?

File: app/routers/categories.py (hunk 11 above)
Comment (ivarshav) [inline app/routers/categories.py]: Consider using mocks instead

File: tests/test_categories.py [hunk 26]
```diff
@@ -9,31 +10,31 @@
...
     @staticmethod
-    def test_creating_new_category(client, user):
-        response = client.post("/categories/",
-                               json={"user_id": user.id, "name": "Foo",
-                                     "color": "eecc11"})
+    def test_creating_new_category(session, client, user):
+        response = client.post("/for_categories_test",
+                               data={"user_id": user.id,
+                                     "category_name": "Foo",
+                                     "chosen_color": "eecc11"})
```
Comment (ivarshav) [inline tests/test_categories.py]: You can add another test with wrong color format

File: tests/test_categories.py [hunk 27]
```diff
@@ -9,31 +10,31 @@
...
     def test_creating_not_unique_category_failed(client, sender, category):
-        response = client.post("/categories/", json={"user_id": sender.id,
-                                                     "name": "Guitar Lesson",
-                                                     "color": "121212"})
-        assert response.status_code == status.HTTP_400_BAD_REQUEST
-        assert TestCategories.CATEGORY_ALREADY_EXISTS_MSG in \
-               response.json()["detail"]
+        response = client.post("/for_categories_test",
+                               data={"category_name": "Guitar Lesson",
+                                     "chosen_color": "121212",
```
Comment (ivarshav) [inline tests/test_categories.py]: I think that the format you expect for colors is "#121212". Is that true?

File: app/templates/event/partials/edit_event_details_tab.html [hunk 28]
```diff
@@ -33,7 +33,19 @@
...
+        <select name="the_categories">
+            {% if categories_list[0] is defined %}
+                <option style="color:{{categories_list[0]['color']}}" value="{{categories_list[0]['name']}}" selected>{{categories_list[0]['name']}}</option>
```
Comment (ivarshav) [inline app/templates/event/partials/edit_event_details_tab.html]: Pay attention that wrong format can cause here lots of problems..

The format you chose is "#aabbcc", right?
I can add validation in BE, please consider FE...

Comment (ivarshav) [reply app/templates/event/partials/edit_event_details_tab.html]: I started to add a validation in BE side, and I think we have some issue -
`#` has a special meaning in URLs (it breaks tests)

So, I suggest that FE will send the color in **"aabbcc"** format, and that's what we will store in the db.
It means that here you need to add `#` before, something like:
```
style="color:#{{categories_list[0]['color']}}"
```
Is that ok with you?

Comment (ivarshav) [reply app/templates/event/partials/edit_event_details_tab.html]: Validation in BE: https://github.com/PythonFreeCourse/calendar/pull/251

File: app/routers/categories.py [hunk 29]
```diff
@@ -35,7 +39,7 @@ class Config:
 
 
 # TODO(issue#29): get current user_id from session
-@router.get("/", include_in_schema=False)
+@router.get("/by_parameters", include_in_schema=False)
```
Comment (yammesicka) [inline app/routers/categories.py]: Maybe `/user`?

File: app/templates/categories.html [hunk 30]
```diff
@@ -0,0 +1,36 @@
...
+{% block head %}
+{{ super() }}
+<link href="{{ url_for('static', path='/categories_style.css') }}" rel="stylesheet">
+<!-- https://github.com/joshbuchea/HEAD -->
+<link
+      rel="stylesheet"
+      href="https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css"
+      integrity="sha384-JcKb8q3iqJ61gNV9KGb8thSsNjpSL0n8PARn9HuZOnIxN0hoP+VmmDGMN5t9UJ0Z"
+      crossorigin="anonymous"
+    />
```
Comment (yammesicka) [inline app/templates/categories.html]: We already use bootstrap 5 :)

File: app/templates/categories.html [hunk 31]
```diff
@@ -0,0 +1,36 @@
...
+<h1>
+  Here you can create your unique categories and choose your favorite color
+</h1>
+
+<form action="/categories" method="POST" required>
+  <p>Name your category:</p>
+  <p><input type="text" id ="name" name="name"></p>
+  <p>Choose your favorite color:</p>
+  <p><input type="color"  class="custom" id ="color", name="color"></p>
+  <p><input type="submit" class="btn btn-success" value="submit" /></p>
```
Comment (yammesicka) [inline app/templates/categories.html]: Use button element insead

File: app/templates/categories.html [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+<h1>It's time to make some decisions</h1>
+<h1>
+  Here you can create your unique categories and choose your favorite color
+</h1>
+
+<form action="/categories" method="POST" required>
+  <p>Name your category:</p>
+  <p><input type="text" id ="name" name="name"></p>
+  <p>Choose your favorite color:</p>
+  <p><input type="color"  class="custom" id ="color", name="color"></p>
```
Comment (yammesicka) [inline app/templates/categories.html]: Add label elements

File: app/templates/event/partials/edit_event_details_tab.html [hunk 33]
```diff
@@ -58,6 +58,18 @@
...
       </select>
    </div>
+   <div class="user_categories">
+      <select name="the_categories">
+          {% if categories_list[0] is defined %}
+              <option style="color:#{{categories_list[0]['color']}}" value="{{categories_list[0]['name']}}" selected>{{categories_list[0]['name']}}</option>
+          {% endif %}
+          {% if categories_list[1] is defined %}
+              {% for category in categories_list[1:] %}
+                  <option style="color:#{{category['color']}}" value="{{category['name']}}" selected>{{category['name']}}</option>
```
Comment (yammesicka) [inline app/templates/event/partials/edit_event_details_tab.html]: Use Jinja's `|safe` feature

File: app/templates/event/partials/edit_event_details_tab.html [hunk 34]
```diff
@@ -58,6 +58,18 @@
          <option value="blue">Blue</option>
       </select>
    </div>
+   <div class="user_categories">
+      <select name="the_categories">
```
Comment (yammesicka) [inline app/templates/event/partials/edit_event_details_tab.html]: Add label

File: app/routers/categories.py [hunk 35]
```diff
@@ -46,39 +49,48 @@ def get_categories(request: Request,
...
                                    f"expected format.")
     try:
-        cat = Category.create(db_sess,
-                              name=category.name,
-                              color=category.color,
-                              user_id=category.user_id)
+        Category.create(db_sess,
+                        name=name,
+                        color=color,
+                        user_id=user_id)
```
Comment (yammesicka) [inline app/routers/categories.py]: Can be in 1 line

File: app/routers/categories.py [hunk 36]
```diff
@@ -46,39 +49,48 @@ def get_categories(request: Request,
...
-                            detail=f"category is already exists for "
-                                   f"user {category.user_id}.")
-    else:
-        return {"category": cat.to_dict()}
+        message = "Category already exists"
+        return templates.TemplateResponse("categories.html", {
+            "request": request,
+            "message": message,
+            "name": name,
+            "color": color,
```
Comment (yammesicka) [inline app/routers/categories.py]: Maybe we should create the dictionary in another function? (It takes lot of space)

**Review by yammesicka — CHANGES_REQUESTED** [review]
Great work! Just a little more :)

File: app/routers/categories.py (hunk 29 above)
Comment (ADI-projects30) [reply app/routers/categories.py]: all the notes have been fixed, ready for merge :)

File: app/templates/event/partials/edit_event_details_tab.html [hunk 37]
```diff
@@ -58,6 +58,19 @@
...
         </div>
+        <div class="user_categories">
+            <label for="categories">Categories</label>
+            <select name="the_categories">
+               {% if categories_list[0] is defined %}
+                  <option style="color:#{{categories_list[0]['color'] |safe }}" value="{{categories_list[0]['name'] |safe }}" selected>{{categories_list[0]['name'] |safe }}</option>
+               {% endif %}
+               {% if categories_list[1] is defined %}
+                  {% for category in categories_list[1:] %}
+                        <option style="color:#{{category['color'] |safe }}" value="{{category['name'] |safe }}" selected>{{category['name'] |safe }}</option>
```
Comment (ivarshav) [inline app/templates/event/partials/edit_event_details_tab.html]: I think selected is redundant here?

File: tests/test_categories.py [hunk 38]
```diff
@@ -20,28 +22,31 @@ def test_get_categories_logic_succeeded(session, user, category):
...
-        assert TestCategories.CATEGORY_ALREADY_EXISTS_MSG in \
-               response.json()["detail"]
+    def test_create_not_unique_category_failed(categories_test_client, sender,
+                                               category):
+        CATEGORY_ALREADY_EXISTS = {"name": "Guitar Lesson",
+                                   "color": "121212",
+                                   "user_id": sender.id}
+        response = categories_test_client.post("/categories/",
+                                               data=CATEGORY_ALREADY_EXISTS)
+        print(response.content)
```
Comment (ivarshav) [inline tests/test_categories.py]: remove print :)

File: tests/test_categories.py [hunk 39]
```diff
@@ -91,10 +99,9 @@ def test_get_category_by_color(client, sender, category):
             ("name", "Guitar Lesson"), ("id", category.id)}
 
     @staticmethod
-    def test_get_category_bad_request(client):
-        response = client.get("/categories/")
```
Comment (ivarshav) [inline tests/test_categories.py:95]: This should be probably `/categories/user` now, could you keep it?

**Comment by ivarshav** [comment]
Thank you for the fixes! Great work :)

File: app/templates/event/partials/edit_event_details_tab.html (hunk 34 above)
Comment (ADI-projects30) [reply app/templates/event/partials/edit_event_details_tab.html]: Noticed the file was moved to templates/partials/calendar/event so I did what you asked me over there.
All the notes have been fixed, ready for merge :)

File: app/templates/event/partials/edit_event_details_tab.html (hunk 33 above)
Comment (ADI-projects30) [reply app/templates/event/partials/edit_event_details_tab.html]: fix :)

File: app/routers/categories.py (hunk 35 above)
Comment (ADI-projects30) [reply app/routers/categories.py]: fix :)

File: app/routers/categories.py (hunk 36 above)
Comment (ADI-projects30) [reply app/routers/categories.py]: fix :)

File: app/templates/categories.html (hunk 32 above)
Comment (ADI-projects30) [reply app/templates/categories.html]: fix :)

File: app/templates/categories.html (hunk 31 above)
Comment (ADI-projects30) [reply app/templates/categories.html]: fix :)

File: app/templates/categories.html (hunk 30 above)
Comment (ADI-projects30) [reply app/templates/categories.html]: fix :)
